frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/contextmenu/PushSelectedBranchAction.java

```diff
@@ -0,0 +1,98 @@
+package com.virtuslab.gitmachete.frontend.actions.contextmenu;
+
+import static io.vavr.API.$;
+import static io.vavr.API.Case;
+import static io.vavr.API.Match;
+
+import com.intellij.openapi.actionSystem.AnActionEvent;
+import com.intellij.openapi.actionSystem.CommonDataKeys;
+import com.intellij.openapi.project.Project;
+import git4idea.repo.GitRepository;
+import io.vavr.control.Option;
+import org.checkerframework.checker.guieffect.qual.UIEffect;
+
+import com.virtuslab.gitmachete.backend.api.SyncToRemoteStatus;
+import com.virtuslab.gitmachete.frontend.actions.common.ActionUtils;
+import com.virtuslab.gitmachete.frontend.actions.common.BasePushBranchAction;
+import com.virtuslab.gitmachete.frontend.datakeys.DataKeys;
+import com.virtuslab.logger.IPrefixedLambdaLogger;
+import com.virtuslab.logger.PrefixedLambdaLoggerFactory;
+
+/**
+ * Expects DataKeys:
+ * <ul>
+ *  <li>{@link DataKeys#KEY_SELECTED_VCS_REPOSITORY}</li>
+ *  <li>{@link DataKeys#KEY_SELECTED_BRANCH_NAME}</li>
+ *  <li>{@link CommonDataKeys#PROJECT}</li>
+ * </ul>
+ */
+public class PushSelectedBranchAction extends BasePushBranchAction {
+  private static final IPrefixedLambdaLogger LOG = PrefixedLambdaLoggerFactory.getLogger("frontendActions");
+
+  @Override
+  @UIEffect
+  public void update(AnActionEvent anActionEvent) {
+    super.update(anActionEvent);
+
+    Option<String> selectedBranchName = ActionUtils.getSelectedBranchName(anActionEvent);
+
+    if (selectedBranchName.isDefined()) {
+      Option<SyncToRemoteStatus> syncToRemoteStatus = ActionUtils.getGitMacheteRepository(anActionEvent)
+          .flatMap(repo -> repo.getBranchByName(selectedBranchName.get()))
+          .flatMap(branch -> Option.of(branch.getSyncToRemoteStatus()));
+
+      if (syncToRemoteStatus.isDefined()) {
+        SyncToRemoteStatus.Relation relation = syncToRemoteStatus.get().getRelation();
+        boolean isEnabled = PUSH_ENABLING_STATUSES.contains(relation);
+
+        if (isEnabled) {
+          Option<Boolean> isSelectedEqualCurrent = ActionUtils.getGitMacheteRepository(anActionEvent)
+              .flatMap(repo -> Option.of(repo.getCurrentBranchIfManaged()))
```

Comment:
Redundant `Option.of`, you have an `Option<Option<...>>` here which is almost always a smell ;)

Reply:
good point!